include/dice/template-library/iter_to_range.hpp

```diff
@@ -0,0 +1,175 @@
+#ifndef DICE_TEMPLATELIBRARY_ITERTORANGE_HPP
+#define DICE_TEMPLATELIBRARY_ITERTORANGE_HPP
+
+#include <cassert>
+#include <cstddef>
+#include <functional>
+#include <iterator>
+#include <optional>
+#include <type_traits>
+#include <utility>
+
+namespace dice::template_library {
+
+	/**
+	 * Wrapper to make a C++-style iterator out of a rust-style iterator.
+	 *
+	 * @tparam Iter base rust-style iterator
+	 *
+	 * Requirements for Iter (the following things must be at least protected):
+	 *		typename Iter::value_type;
+	 *		{ Iter::next() } -> std::same_as<std::optional<typename Iter::value_type>>;
+	 *
+	 * We are not using a concept here, because concepts require things to be public
+	 */
+	template<typename Iter>
+	struct iter_to_iter : Iter {
+		using base_iterator = Iter;
+		using sentinel = std::default_sentinel_t;
+		using value_type = typename Iter::value_type;
+		using reference = value_type const &;
+		using pointer = value_type const *;
+		using difference_type = std::ptrdiff_t;
+		using iterator_category = std::input_iterator_tag;
+
+	private:
+		std::optional<value_type> cur_;
+		std::optional<value_type> peeked_;
+
+		void advance() {
+			if (peeked_.has_value()) [[unlikely]] {
+				// fast path, we have already peaked the value
+				cur_ = std::exchange(peeked_, std::nullopt);
+			} else {
+				cur_ = this->next();
+			}
+		}
+
+	public:
+		template<typename ...Args>
+		explicit iter_to_iter(Args &&...args) : Iter{std::forward<Args>(args)...},
+												cur_{this->next()} {
+		}
+
+		[[nodiscard]] reference operator*() const noexcept {
+			assert(cur_.has_value());
+			return *cur_;
+		}
+
+		[[nodiscard]] pointer operator->() const noexcept {
+			assert(cur_.has_value());
+			return &*cur_;
+		}
+
+		/**
+		 * Access the current value of the iterator mutably.
```

Comment:
I am not sure what this implies. With this method we could change `cur_`. But since it is a non-pointer field that doesn't change the underlying data structure.

Reply:
It allows moving values out of the iterator